src/editor.rs

```diff
@@ -54,8 +60,51 @@ impl Editor {
         self.selection
     }
 
+    fn insert_line(&mut self, line_no: usize, start_pos: usize) {
+        // Re-map lines that will be affected by this change.
+        for val in self.id_map.values_mut() {
+            if *val > line_no {
+                *val += 1;
+            }
+        }
+        // Update the position map.
+        self.id_map.insert(self.next_id, line_no + 1);
+
+        let the_line = self.lines.lines().at_unkeyed(line_no);
+        let new_text = if start_pos < the_line.read_untracked().text().len()
+            && !the_line
+                .read_untracked()
+                .text()
+                .starts_with(Self::COSMETIC_SPACE)
```

Comment:
I think it would be better if everything having to do with the display of empty lines remained private to the line (or at least private to somewhere). Only one module should be thinking about the existence of the cosmetic space...